specification/dsdl/conventions.tex

```diff
@@ -0,0 +1,115 @@
+
+Every data type definition file should begin with a header comment that provides an exhaustive description
+of the data type, its purpose, semantics, usage patterns, any related data exchange patterns,
+assumptions, constraints, and all other information that may be necessary or generally useful for the usage of the
+data type definition.
+
+Every attribute of the data type definition, and especially every field attribute of it,
+should have an associated comment explaining the purpose of the attribute, its semantics, usage patterns,
+assumptions, constraints, and any other pertinent information.
+Exception applies to attributes supplied with sufficiently descriptive and unambiguous names.
+
+% Field comment placement https://forum.uavcan.org/t/dsdl-documentation-comments/407
+
+\subsection{Optional value representation}
+
+Data structures may include optional field attributes that are not necessarily always populated.
+
+The recommended approach for representing optional field attributes
+is to use variable-length arrays with the capacity of one element,
+prefixed with padding bits as necessary to retain byte alignment.
+
+Alternatively, such one-element variable-length arrays can be replaced with two-field unions,
+where the first field is empty and the second field contains the desired optional value.
+The described layout is bit-compatible and semantically compatible with the one-element array described above,
+provided that the field attributes are not swapped.
+
+Floating-point-typed field attributes may be assigned the value of not-a-number (NaN) per IEEE 754
+to indicate that the value is not specified;
+however, this pattern is discouraged because some nodes may not support IEEE 754 NaN values internally,
+the value would still have to be transferred over the bus even if not populated,
+and special case values undermine type safety.
+
+\begin{remark}[breakable]
+    Array-based optional field:
+
+    \begin{minted}{python}
+        void7                           # It is recommended to ensure byte alignment.
+        MyType[<=1] optional_field
+    \end{minted}
+
+    Union-based optional field:
+
+    \begin{minted}{python}
+        @union                          # The implicit tag is one bit long.
+        uavcan.primitive.Empty none     # Represents lack of value, unpopulated field.
+        MyType some                     # The field of interest; field ordering is important.
+    \end{minted}
+
+    The defined above union can be used as follows (suppose it is named \verb|MaybeMyType|):
+
+    \begin{minted}{python}
+        void7                           # It is recommended to ensure byte alignment.
+        MaybeMyType optional_field
+    \end{minted}
+
+    The shown approaches are mutually bit-compatible and semantically compatible.
+\end{remark}
+
+\subsection{Bit flag representation}
+
+The recommended approach to defining a set of bit flags is to dedicate a \verb|bool|-typed field attribute for each.
```

Comment:
Similar comment here as above: I think we have a normative specification that "bit flags are not, intrinsically, supported" and a non-normative section about how they can be emulated.

Reply:
> I think we have a normative specification that "bit flags are not, intrinsically, supported"

But they are supported, are they not? We have `bool` which is encoded as a single bit. Should it say instead "Bit flags may be represented as fields of type `bool`" in the normative section? I don't see any issues with the current wording but I'm open to changing it.

Reply:
If this is normative then we aren't recommending we are specifying. Which is it for bit flags?

Reply:
There is a normative section that defines fields of type `bool`. Then there is a non-normative section (the one we're commenting under) that says that if you want a set of bit flags, you are advised to represent them in DSDL as a set of fields of type `bool` rather than a single field of type `uint*`.

Reply:
okay.